### PR TITLE
Esphome native and NDEF

### LINKED ISSUE #41: Expose Ndef to act as a tag
PN532 should be able to expose the NDEF, this would make the tag reader a configurable tag with i.e. 
wifi settings ... 
or even maybe a dynamic game further down the line

### PR DESCRIPTION
**Pull request by jesserockz**
This PR uses native esphome components that have been recently developed and requires that you use ESPHome 1.16.0.

closes #41 

## What is this?

This PR switches to use the ESPHome native version of `pn532_i2c` and also as an extension to that NDEF reading and writing.

### CODE REVIEW

**Comment by RichieFrame** [comment]
Testing this now with a custom yaml, working much better than the non-native i2c version, which would only read a tag if it was on the reader during boot, then not read anymore once it was removed (#54 on steroids).

Thanks to a bad reader board this took 2 days to get working, with the non-native code, there was no error, this actually reports the board bad:

[01:39:46][C][pn532:015]: Setting up PN532...
[01:39:46][W][i2c:076]: Unknown transmit error 4 for address 0x24
[01:39:46][W][i2c:086]: Requesting 1 bytes from 0x24 failed!
[01:39:46][E][pn532:019]: Error sending version command
[01:39:46][E][component:092]: Component was marked as failed.

**Comment by RichieFrame** [comment]
ESP32 test did not go so hot, it did work using SPI, but this is what I get with native i2c:

```
[03:21:49][C][pn532:015]: Setting up PN532...
[03:21:49][D][pn532:030]: Found chip PN532
[03:21:49][D][pn532:031]: Firmware ver. 1.6

[03:21:49][W][i2c:086]: Requesting 1 bytes from 0x24 failed!
[03:21:49][W][i2c:086]: Requesting 1 bytes from 0x24 failed!
[03:21:49][W][i2c:086]: Requesting 1 bytes from 0x24 failed!
[03:21:49][E][esp32-hal-i2c.c:1426] i2cCheckLineState(): Bus Invalid State, TwoWire() Can't init sda=1, scl=0
[03:21:49]
[03:21:49][E][esp32-hal-i2c.c:1426] i2cCheckLineState(): Bus Invalid State, TwoWire() Can't init sda=1, scl=0
[03:21:49][W][i2c:086]: Requesting 1 bytes from 0x24 failed!
[03:21:50][E][esp32-hal-i2c.c:1426] i2cCheckLineState(): Bus Invalid State, TwoWire() Can't init sda=1, scl=0
[03:21:50][W][i2c:086]: Requesting 1 bytes from 0x24 failed!

[03:21:50][E][esp32-hal-i2c.c:1426] i2cCheckLineState(): Bus Invalid State, TwoWire() Can't init sda=1, scl=0
[03:21:50][W][i2c:076]: Unknown transmit error 5 for address 0x24
[03:21:50][E][esp32-hal-i2c.c:1426] i2cCheckLineState(): Bus Invalid State, TwoWire() Can't init sda=1, scl=0
[03:21:50][W][i2c:086]: Requesting 1 bytes from 0x24 failed!
[03:21:50][E][esp32-hal-i2c.c:1426] i2cCheckLineState(): Bus Invalid State, TwoWire() Can't init sda=1, scl=0
[03:21:50][W][i2c:086]: Requesting 1 bytes from 0x24 failed!
[03:21:50][W][pn532:087]: Requesting tag read failed!
```

With that final block of 7 lines repeating every second after until i kill the serial connection

I am using GPIO 21 and 22 for SDA and SCL, tested with 5V and 3.3V power, 100KHz and 200KHz
Not sure if that is a board issue or not, those were the same pins I was using for SPI CLK and MISO and they worked

**Comment by jesserockz** [comment]
Oh man, thats going to be hard to look into as I do not have an esp32 to test with...
A very quick google brings this: espressif/arduino-esp32#3079 Might need pullups on the esp32?

**Comment by RichieFrame** [comment]
Interesting, when I get time I will try that. I have 2.2K and 1K that I can put in series to get pretty close, or I can do 22K x 2 + 4.7K and that will get me to 3.293K

**Comment by RichieFrame** [comment]
Actually, I just went through the NXP electrical bus specs for I2C and the Espressif I2C driver parameters... there are pullups on those pins, but they are not sufficient for I2C, and they have to be enabled in the driver, and tuned to both the bus frequency and capacitance.

I calculated a minimum of 1K and a max of 3.6K @ 400KHz. 4.7K @ 200KHz is confirmed working now on a breadboard.

It might also be recommended to add an inline current limiting resistor, for the device, I cant find a board schematic but I can see resistors on the SPI bus pins only

File: tagreader.yaml [hunk 1]
```diff
@@ -60,24 +60,98 @@ api:
...
+      - while:
+          condition:
+            lambda: 'return id(continous_writing_mode);'
+          then:
+          - lambda: |-
+              static const char alphanum[] = "0123456789abcdef";
+              std::string uri = "https://www.home-assistant.io/tag/";
+              for (int i = 0; i < 8; i++)
+                uri += alphanum[random_uint32() % (sizeof(alphanum) - 1)];
+              uri += "-";
```
Comment (balloob) [inline tagreader.yaml:77]: This is optional. For HA any random string works. I guess it makes it more readable / mimics what iOS and Android write.

**Review by balloob — APPROVED** [review]

**Comment by balloob** [comment]
Do we need to update the README to specify the version of ESPHome that is needed?

**Comment by adonno** [comment]
> Do we need to update the README to specify the version of ESPHome that is needed?

Wouldn't hurt ?

**Comment by balloob** [comment]
This is not about hurting. This is about if the native integration requires a minimum version of ESPHome.

**Comment by RichieFrame** [comment]
I would suggest increasing the polling rate, with the old SPI it was always on and responded instantly, now there can be a delay that exceeds the amount of time someone is willing to hold when 'tapping' a card to the reader.

I would also suggest moving the polling rate control from static code in the driver python file and into the config yaml code, having a default of 500ms unless changed, (with 0 for no sleep), I set it to 250ms to test, and it is getting warmer (less than +10c), but not hot like SPI did (+30c)

**Comment by jesserockz** [comment]
Hi @RichieFrame 
You can change the polling rate by using the below
```yaml
pn532_i2c:
  update_interval: 500ms
  ...
```

**Comment by RichieFrame** [comment]
and that works!

I did do more testing, 350ms seems to be the best balance of power/responsiveness so far, temp is about +6.5c over ambient

**Comment by RichieFrame** [comment]
To confirm, the code is not in place to write MIME records correct? I assume it can read them with the correct esphome code?

What about checking the tag capacity prior to writing data? What happens if we try to write more data than the tag can hold?

**Comment by jesserockz** [comment]
At the moment, there are helpers in the `NdefMessage` in ESPHome to `message.add_text_record` and `message.add_uri_record` but you can always make a `new NdefRecord(...)` with the data that you want and just call `message.add_record(record)`.

Currently there is no capacity check for a mifare classic card. The ultralight/type 2 cards do check the capacity first as they are known to be smaller. Im not sure what happens if you do try to write more

**Comment by dcjona** [comment]
hi, is it possible to install your branch on docker?

**Pull request merged**